File: Sources/i18nCore/ExcelParser.swift
import Foundation

/// Parses CSV content downloaded from Google Sheets into `LocalizationData`.
///
/// Expected column layout:
/// | key | platform | default | en | my | zh | ... |
public enum ExcelParser {
    private static let requiredHeaders = ["key", "platform"]
    private static let defaultColumn = "default"

    public static func parse(_ csvContent: String) throws -> LocalizationData {
        let lines = csvContent.components(separatedBy: .newlines).filter { !$0.trimmingCharacters(in: .whitespaces).isEmpty }

        guard !lines.isEmpty else {
            return LocalizationData(defaultLocale: defaultColumn, additionalLocales: [], entries: [])
        }

        let headers = parseCsvLine(lines[0])
        try validateHeaders(headers)

        let defaultIndex = headers.firstIndex(of: defaultColumn) ?? -1
        let additionalLocales = headers.filter { $0 != defaultColumn && !requiredHeaders.contains($0) }

        let entries = lines.dropFirst().compactMap { line -> LocalizationEntry? in
            parseEntry(parseCsvLine(line), defaultIndex: defaultIndex, additionalLocales: additionalLocales)
        }

        return LocalizationData(defaultLocale: defaultColumn, additionalLocales: additionalLocales, entries: entries)
    }

    private static func validateHeaders(_ headers: [String]) throws {
        for required in requiredHeaders {
            guard headers.contains(required) else {
                throw I18nError.missingHeader(required)
            }
        }
        guard headers.contains(defaultColumn) else {
            throw I18nError.missingHeader(defaultColumn)
        }
    }

    private static func parseCsvLine(_ line: String) -> [String] {
        var result: [String] = []
        var remaining = line[line.startIndex...]

        while !remaining.isEmpty {
            if remaining.first == "\"" {
                remaining = remaining.dropFirst()
                if let endQuote = remaining.firstIndex(of: "\"") {
                    result.append(String(remaining[..<endQuote]))
                    remaining = remaining[remaining.index(after: endQuote)...]
                    if remaining.first == "," { remaining = remaining.dropFirst() }
                } else {
                    result.append(String(remaining))
                    remaining = remaining[remaining.endIndex...]
                }
            } else if let comma = remaining.firstIndex(of: ",") {
                result.append(String(remaining[..<comma]))
                remaining = remaining[remaining.index(after: comma)...]
            } else {
                result.append(String(remaining))
                remaining = remaining[remaining.endIndex...]
            }
        }
        return result
    }

    private static func parseEntry(_ values: [String], defaultIndex: Int, additionalLocales: [String]) -> LocalizationEntry? {
        guard !values.isEmpty else { return nil }

        let key = values[0].trimmingCharacters(in: .whitespaces)
        guard !key.isEmpty else { return nil }

        let platform = values.count > 1 ? values[1].trimmingCharacters(in: .whitespaces).lowercased() : ""
        let defaultValue = defaultIndex >= 0 && defaultIndex < values.count
            ? values[defaultIndex].trimmingCharacters(in: .whitespaces)
            : ""

        var translations: [String: String] = [:]
        for (localeIndex, locale) in additionalLocales.enumerated() {
            let valueIndex = 3 + localeIndex
            if valueIndex < values.count {
                translations[locale] = values[valueIndex].trimmingCharacters(in: .whitespaces)
            }
        }

        return LocalizationEntry(key: key, platform: platform, defaultValue: defaultValue, translations: translations)
    }
}

public enum I18nError: Error, LocalizedError {
    case missingHeader(String)
    case invalidSheetUrl(String)
    case downloadFailed(String)
    case configNotFound
    case configInvalid(String)

    public var errorDescription: String? {
        switch self {
        case .missingHeader(let h): return "Missing required CSV header: \(h)"
        case .invalidSheetUrl(let u): return "Invalid Google Sheet URL: \(u)"
        case .downloadFailed(let msg): return "Failed to download sheet: \(msg)"
        case .configNotFound: return "i18n.json not found. Create one in your project root."
        case .configInvalid(let msg): return "i18n.json is invalid: \(msg)"
        }
    }
}

File: plugin/src/main/kotlin/io/github/fcat97/i18n/parser/ExcelParser.kt
package io.github.fcat97.i18n.parser

data class LocalizationEntry(
    val key: String,
    val platform: String,
    val defaultValue: String,
    val translations: Map<String, String>
)

data class LocalizationData(
    val defaultLocale: String,
    val additionalLocales: List<String>,
    val entries: List<LocalizationEntry>
)

object ExcelParser {
    private val REQUIRED_HEADERS = listOf("key", "platform")
    private const val DEFAULT_COLUMN = "default"

    fun parse(csvContent: String): LocalizationData {
        val lines = csvContent.lines().filter { it.isNotBlank() }

        if (lines.isEmpty()) {
            return LocalizationData(DEFAULT_COLUMN, emptyList(), emptyList())
        }

        val headers = parseCsvLine(lines[0])
        validateHeaders(headers)

        val defaultIndex = headers.indexOf(DEFAULT_COLUMN)
        val additionalLocales = headers.filter { it != DEFAULT_COLUMN && it !in REQUIRED_HEADERS }
        val entries = lines.drop(1).mapNotNull { line ->
            parseEntry(parseCsvLine(line), defaultIndex, additionalLocales)
        }

        return LocalizationData(DEFAULT_COLUMN, additionalLocales, entries)
    }

    private fun validateHeaders(headers: List<String>) {
        REQUIRED_HEADERS.forEach { required ->
            if (required !in headers) {
                throw IllegalArgumentException("Missing required header: $required")
            }
        }
        if (DEFAULT_COLUMN !in headers) {
            throw IllegalArgumentException("Missing required header: $DEFAULT_COLUMN")
        }
    }

    private fun parseCsvLine(line: String): List<String> {
        val result = mutableListOf<String>()
        var remaining = line

        while (remaining.isNotEmpty()) {
            if (remaining.startsWith("\"")) {
                remaining = remaining.substring(1)
                val endQuote = remaining.indexOf("\"")
                if (endQuote >= 0) {
                    result.add(remaining.substring(0, endQuote))
                    remaining = if (endQuote + 1 < remaining.length) remaining.substring(endQuote + 1) else ""
                    if (remaining.startsWith(",")) remaining = remaining.substring(1)
                } else {
                    result.add(remaining)
                    remaining = ""
                }
            } else {
                val commaIndex = remaining.indexOf(",")
                if (commaIndex >= 0) {
                    result.add(remaining.substring(0, commaIndex))
                    remaining = remaining.substring(commaIndex + 1)
                } else {
                    result.add(remaining)
                    remaining = ""
                }
            }
        }
        return result
    }

    private fun parseEntry(values: List<String>, defaultIndex: Int, additionalLocales: List<String>): LocalizationEntry? {
        if (values.isEmpty()) return null

        val key = values[0].trim()
        if (key.isEmpty()) return null

        val platform = if (values.size > 1) values[1].trim().lowercase() else ""
        val defaultValue = if (defaultIndex >= 0 && defaultIndex < values.size) values[defaultIndex].trim() else ""

        val translations = mutableMapOf<String, String>()
        additionalLocales.forEachIndexed { localeIndex, locale ->
            val valueIndex = 3 + localeIndex
            if (valueIndex < values.size) {
                translations[locale] = values[valueIndex].trim()
            }
        }

        return LocalizationEntry(key, platform, defaultValue, translations)
    }
}
